src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -23332,6 +23334,361 @@ struct DynamicPadToPad
   }
 };
 
+struct RemoveNoOpsFromWhileLoop
+    : public CheckedOpRewritePattern<stablehlo::WhileOp,
+                                     RemoveNoOpsFromWhileLoop> {
+  using CheckedOpRewritePattern<
+      stablehlo::WhileOp, RemoveNoOpsFromWhileLoop>::CheckedOpRewritePattern;
+
+  LogicalResult matchAndRewriteImpl(stablehlo::WhileOp whileOp,
+                                    PatternRewriter &rewriter) const {
+    auto info = WhileLoopInfo(whileOp);
+    auto computeInfoSuccess = info.computeInfo();
+    if (computeInfoSuccess.failed())
+      return computeInfoSuccess;
+
+    if (!info.isValid() || !info.isConstant())
+      return failure();
+
+    auto &whileBody = whileOp.getBody().front();
+    auto inductionVar = whileBody.getArgument(0);
+
+    bool anyNoOpRemoved = false;
+
+    auto limit = info.getConstantLimit().value();
+    auto start = info.getConstantStart().value();
+    auto step = info.getConstantStep().value();
+
+    auto inductionType = inductionVar.getType();
+    unsigned bitWidth = 64;
+    if (auto tensorType = dyn_cast<RankedTensorType>(inductionType)) {
+      if (auto intType = dyn_cast<IntegerType>(tensorType.getElementType())) {
+        bitWidth = intType.getWidth();
+      }
+    }
+
+    // Initialize bounds map with induction variable bounds
+    DenseMap<Value, Bounds> boundsMap;
+    if (step > 0) {
+      APInt minBound(bitWidth, start, true);
+      APInt maxBound(bitWidth, limit - 1, true);
```

Comment:
this isn't necessarily correct, since start == limit is a valid case [where the loop body isn't executed]

Reply:
though I suppose if the body is never executed, what is done in the after region doesn't matter anyways

Reply:
I'm not seeing where since if step > limit - start, loop count is still zero [and step only checked for sign]

Reply:
A bit above

```
    if (!info.isValid() || !info.isConstant() ||
        info.getConstantNumIters() <= 0)
      return failure();
```

Reply:
ah ok